dotnet/src/SemanticKernel.Abstractions/Kernel.cs

```
@@ -99,155 +152,148 @@ public CultureInfo Culture
     /// <remarks>
-    /// The KernelBuilder class provides a fluent API for constructing a <see cref="Kernel"/> instance.
+    /// The behavior of this method is not the same as that of <see cref="IServiceProvider.GetService(Type)"/>
+    /// on the exposed <see cref="Services"/>. Rather, it is opinionated view around it. If a <paramref name="serviceId"/>
+    /// is provided, it will attempt to find a service registered with that key. If no <paramref name="serviceId"/>
+    /// is provided, it will attempt to find any service registered, regardless of whether it was registered with
+    /// with a key. If multiple services meet the criteria, it will return one of those registered, but no guarantee
+    /// on exactly which. For certain services, like <see cref="ILoggerFactory"/>, it will also return a default implementation
+    /// if no key was specified and no service was found. If it's able to find the specified service, that service is returned.
+    /// Otherwise, an exception is thrown.
     /// </remarks>
-    public Kernel(
-        IAIServiceProvider aiServiceProvider,
-        IEnumerable<IKernelPlugin>? plugins = null,
-        IAIServiceSelector? serviceSelector = null,
-        IDelegatingHandlerFactory? httpHandlerFactory = null,
-        ILoggerFactory? loggerFactory = null)
+    public T GetService<T>(string? serviceId = null) where T : class
     {
-        Verify.NotNull(aiServiceProvider);
+        T? service = null;
 
-        this.ServiceProvider = aiServiceProvider;
-        this._plugins = plugins is not null ? new KernelPluginCollection(plugins) : null;
-        this._serviceSelector = serviceSelector;
-        this.HttpHandlerFactory = httpHandlerFactory ?? NullHttpHandlerFactory.Instance;
-        this.LoggerFactory = loggerFactory ?? NullLoggerFactory.Instance;
+        if (serviceId is not null)
+        {
+            if (this.Services is IKeyedServiceProvider)
```

Should we throw a better exception if a `serviceId` is provided, but the service provider isn't an `IKeyedServiceProvider`? Right now it would just say `"Service of type '{typeof(T)}' and key '{serviceId}' not registered."`

---

I can tweak the message for that case.